main.py

```diff
@@ -0,0 +1,449 @@
+    heightmap_resolution = args.heightmap_resolution # Meters per pixel of heightmap
+    random_seed = args.random_seed
+    force_cpu = args.force_cpu
+
+    # ------------- Algorithm options -------------
+    method = args.method # 'reactive' (supervised learning) or 'reinforcement' (reinforcement learning ie Q-learning)
+    push_rewards = args.push_rewards if method == 'reinforcement' else None  # Use immediate rewards (from change detection) for pushing?
+    future_reward_discount = args.future_reward_discount
+    experience_replay = args.experience_replay # Use prioritized experience replay?
+    heuristic_bootstrap = args.heuristic_bootstrap # Use handcrafted grasping algorithm when grasping fails too many times in a row?
+    explore_rate_decay = args.explore_rate_decay
+    grasp_only = args.grasp_only
+
+    # -------------- Testing options --------------
+    is_testing = args.is_testing
+    max_test_trials = args.max_test_trials # Maximum number of test runs per case/scenario
+    test_preset_cases = args.test_preset_cases 
+    test_preset_file = os.path.abspath(args.test_preset_file) if test_preset_cases else None
+
+    # ------ Pre-loading and logging options ------
+    load_snapshot = args.load_snapshot # Load pre-trained snapshot of model?
+    snapshot_file = os.path.abspath(args.snapshot_file)  if load_snapshot else None
+    continue_logging = args.continue_logging # Continue logging from previous session
+    logging_directory = os.path.abspath(args.logging_directory) if continue_logging else os.path.abspath('logs')
+    save_visualizations = args.save_visualizations # Save visualizations of FCN predictions? Takes 0.6s per training step if set to True
+
+
+    # Set random seed
+    np.random.seed(random_seed)
+
+    # Initialize pick-and-place system (camera and robot)
+    robot = Robot(is_sim, obj_mesh_dir, num_obj, workspace_limits,
+                  tcp_host_ip, tcp_port, rtc_host_ip, rtc_port,
+                  is_testing, test_preset_cases, test_preset_file)
+
+    # Initialize trainer
+    trainer = Trainer(method, push_rewards, future_reward_discount,
+                      is_testing, load_snapshot, snapshot_file, force_cpu)
+
+    # Initialize data logger
+    logger = Logger(continue_logging, logging_directory)
+    logger.save_camera_info(robot.cam_intrinsics, robot.cam_pose, robot.cam_depth_scale) # Save camera intrinsics and pose
+    logger.save_heightmap_info(workspace_limits, heightmap_resolution) # Save heightmap parameters
+
+    # Find last executed iteration of pre-loaded log, and load execution info and RL variables
+    if continue_logging:
+        trainer.preload(logger.transitions_directory)
+
+    # Initialize variables for heuristic bootstrapping and exploration probability
+    no_change_count = [2, 2] if not is_testing else [0, 0]
```

Comment:
Why is no_change_count = [2,2], what do the 2's represent?

Reply:
Look at the comment below "If heuristic bootstrapping is enabled: if change has not been detected more than 2 times, execute heuristic algorithm to detect grasps/pushes". 

The count is how many times it has changed, so I guess they want to use the heuristic algorithm instead of exploring different states with exploration probability.